Quality-of-Life for Google Colab

woctezuma opened the pull request:
Hello,

I know you don't accept pull requests. However, this could be of interest to others who want to run the code on Google Colab,

To put it in a nutshell, I have ported my changes from https://github.com/NVlabs/stylegan2-ada/pull/6 and a bit more:
- save output images as JPG,
- automatically resume from the latest `.pkl` file with the command-line argument `--resume=latest`,
- automatically set the resume value of `kimg`,
- automatically set the resume value of the augmentation strength,
- allow to **manually** set the resume value of the augmentation strength,
- add config `auto_norp` to replicate the `auto` config without EMA rampup,
- allow to override mapping net depth with the command-line argument `--cfg_map`,
- allow to enforce CIFAR-specific architecture tuning with the command-line argument `--cifar_tune`.

I have tested the training with my changes [in one of my repository](https://github.com/woctezuma/steam-stylegan2-ada-pytorch), in order to fix bugs which I could have introduced. It seems fine.

woctezuma committed:
Save output as JPG instead of PNG

woctezuma committed:
Add utility functions

woctezuma committed:
Resume from the latest pickle

woctezuma committed:
Automatically set the resume value of kimg

woctezuma committed:
Allow to manually set the resume value of the augmentation strength

woctezuma committed:
Add cfg (auto_norp): auto cfg without EMA rampup

woctezuma committed:
Allow to override mapping net depth with --cfg_map

woctezuma committed:
Allow to enforce CIFAR-specific architecture tuning with --cifar_tune

woctezuma committed:
README: add changelog

woctezuma committed:
Automatically set the resume value of augment_p

estan commented:
Thanks for this @woctezuma! This will probably come in handy for me (even if I'm not running on colab).

Some questions:

- In your example training in https://github.com/woctezuma/steam-stylegan2-ada-pytorch/blob/main/training.ipynb, you don't use any EMA ramp-up even for the initial training. Do you do this because in that example, you are doing transfer learning from ffhq256?
- In my case, I would like to do from scratch training on my own dataset, so I would use `--cfg auto` for the initial training, but then `--cfg auto_norp` when I resume?
- Or is the `auto_norp` config you added not related to resume at all? (and in the example you simply did not want to use EMA ramp-up at all? Sorry, I don't quite know what EMA ramp-up does, and want to use the right params for my initial training and for my resumptions.

woctezuma commented:
I added `auto_norp` to let people disable the Exponential Moving Average (EMA) ramp-up.
The ramp-up controls the update of `G_ema` as shown below:

https://github.com/NVlabs/stylegan2-ada-pytorch/blob/d4b2afe9c27e3c305b721bc886d2cb5229458eba/training/training_loop.py#L104-L105

https://github.com/NVlabs/stylegan2-ada-pytorch/blob/d4b2afe9c27e3c305b721bc886d2cb5229458eba/training/training_loop.py#L296-L303

---

> Do you do this because in that example, you are doing transfer learning from ffhq256?

There are 2 observations to be made in the code released by Nvidia:

1) EMA ramp-up is disabled when "resuming", which actually means transfer learning in the original code:

https://github.com/NVlabs/stylegan2-ada-pytorch/blob/d4b2afe9c27e3c305b721bc886d2cb5229458eba/train.py#L313-L315

2) EMA ramp-up is disabled for every config except `auto` and `cifar`:

https://github.com/NVlabs/stylegan2-ada-pytorch/blob/d4b2afe9c27e3c305b721bc886d2cb5229458eba/train.py#L154-L161

https://github.com/NVlabs/stylegan2-ada-pytorch/blob/d4b2afe9c27e3c305b721bc886d2cb5229458eba/train.py#L192-L193

---

> In my case, I would like to do from scratch training on my own dataset, so I would use --cfg auto for the initial training, but then --cfg auto_norp when I resume?

This is your decision here. Beware that parameters automatically set are not great. For instance:
- you would not find the optimal parameters used in the paper's experiments if you used `auto`,
- I would suggest to try to manually adjust `gamma`, e.g. try to increase it if you have a small dataset and some computation resources. cf. the README which I quote below:

> `--gamma=10` overrides R1 gamma.
> We recommend trying a couple of different values [for gamma] for each new dataset.

---

> Or is the auto_norp config you added not related to resume at all? (and in the example you simply did not want to use EMA ramp-up at all?

A bit of both as mentioned in my reply to your first question: EMA ramp-up is disabled 1) when fine-tuning or 2) when running the paper's experiments (except for the CIFAR one).

estan commented:
Thanks a lot for the explanation and tips @woctezuma.

thusinh1969 commented:
It costed me 10,000 USD already, still NOT CONVERGENCE !!!

woctezuma commented:
Don't spend so much money! You are not guaranteed that it would work! It depends on data, parameters, etc.

ekerstein commented:
This branch is very helpful, thank you 🙏

jannehellsten committed:
pytorch 1.11 support: don't use conv2d_gradfix on v1.11, port grid_sample_gradfix to the new API

thanks @timothybrooks for the fix!

for #145

jannehellsten committed:
Fix custom ops bug for pytorch 1.12 and onwards

Adapt to newer _jit_get_operation API that changed in
https://github.com/pytorch/pytorch/pull/76814

for #188, #193

woctezuma committed:
Support newer versions of PyTorch (v1.1X and v2)
